File: BoardGameGeek/src/com/boardgamegeek/provider/GamesIdPublishersProvider.java
package com.boardgamegeek.provider;

import android.content.ContentValues;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;

import com.boardgamegeek.provider.BggContract.Games;
import com.boardgamegeek.provider.BggContract.Publishers;
import com.boardgamegeek.provider.BggContract.SyncColumns;
import com.boardgamegeek.provider.BggDatabase.GamesPublishers;
import com.boardgamegeek.provider.BggDatabase.Tables;
import com.boardgamegeek.util.SelectionBuilder;

public class GamesIdPublishersProvider extends BaseProvider {
	private static final String TABLE = Tables.GAMES_PUBLISHERS;

	@Override
	protected SelectionBuilder buildExpandedSelection(Uri uri) {
		final int gameId = Games.getGameId(uri);
		return new SelectionBuilder().table(Tables.GAMES_PUBLISHERS_JOIN_PUBLISHERS)
				.mapToTable(Publishers._ID, Tables.PUBLISHERS).mapToTable(Publishers.PUBLISHER_ID, Tables.PUBLISHERS)
				.mapToTable(SyncColumns.UPDATED, Tables.PUBLISHERS)
				.whereEquals(Tables.GAMES_PUBLISHERS + "." + GamesPublishers.GAME_ID, gameId);
	}

	@Override
	protected SelectionBuilder buildSimpleSelection(Uri uri) {
		int gameId = Games.getGameId(uri);
		return new SelectionBuilder().table(TABLE).whereEquals(GamesPublishers.GAME_ID, gameId);
	}

	@Override
	protected String getDefaultSortOrder() {
		return Publishers.DEFAULT_SORT;
	}

	@Override
	protected String getPath() {
		return "games/#/publishers";
	}

	@Override
	protected String getType(Uri uri) {
		return Publishers.CONTENT_TYPE;
	}

	@Override
	protected Uri insert(SQLiteDatabase db, Uri uri, ContentValues values) {
		long rowId = db.insertOrThrow(TABLE, null, values);
		return Games.buildPublisherUri(rowId);
	}
}

File: app/src/main/java/com/boardgamegeek/provider/GamesIdPublishersProvider.kt
package com.boardgamegeek.provider

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.net.Uri
import android.provider.BaseColumns
import com.boardgamegeek.provider.BggContract.*
import com.boardgamegeek.provider.BggContract.Companion.PATH_GAMES
import com.boardgamegeek.provider.BggContract.Companion.PATH_PUBLISHERS
import com.boardgamegeek.provider.BggDatabase.GamesPublishers.GAME_ID
import com.boardgamegeek.provider.BggDatabase.Tables

class GamesIdPublishersProvider : BaseProvider() {
    override fun getType(uri: Uri) = Publishers.CONTENT_TYPE

    override val path = "$PATH_GAMES/#/$PATH_PUBLISHERS"

    override val defaultSortOrder = Publishers.DEFAULT_SORT

    public override fun buildSimpleSelection(uri: Uri): SelectionBuilder {
        val gameId = Games.getGameId(uri)
        return SelectionBuilder().table(Tables.GAMES_PUBLISHERS).whereEquals(GAME_ID, gameId)
    }

    override fun buildExpandedSelection(uri: Uri): SelectionBuilder {
        val gameId = Games.getGameId(uri)
        return SelectionBuilder()
            .table(Tables.GAMES_PUBLISHERS_JOIN_PUBLISHERS)
            .mapToTable(BaseColumns._ID, Tables.PUBLISHERS)
            .mapToTable(Publishers.Columns.PUBLISHER_ID, Tables.PUBLISHERS)
            .mapToTable(Publishers.Columns.UPDATED, Tables.PUBLISHERS)
            .whereEquals("${Tables.GAMES_PUBLISHERS}.$GAME_ID", gameId)
    }

    override fun insert(context: Context, db: SQLiteDatabase, uri: Uri, values: ContentValues): Uri {
        values.put(GAME_ID, Games.getGameId(uri))
        val rowId = db.insertOrThrow(Tables.GAMES_PUBLISHERS, null, values)
        return Games.buildPublisherUri(rowId)
    }
}
